### PR TITLE
Adding initial Host load check API support

### PR DESCRIPTION
**Pull request by mathewc**

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs [hunk 1]
```diff
@@ -107,13 +110,20 @@ public HostStatus GetHostStatus()
...
+                }
+
+                return Ok(status);
             }
             else
             {
                 // for Anonymous requests, we don't return any
                 // detailed info
-                return status;
+                return Ok();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: For Anonymous requests the status endpoint can serve as a simple ping

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs [hunk 2]
```diff
@@ -87,18 +88,20 @@ public FunctionStatus GetFunctionStatus(string name)
...
+        public IHttpActionResult GetHostStatus()
         {
-            var status = new HostStatus();
-
             // based on the authorization level we determine
             // the additional level of detail to return
             var authorizationLevel = Request.GetAuthorizationLevel();
-            if (authorizationLevel == AuthorizationLevel.Admin)
+            if (authorizationLevel == AuthorizationLevel.Admin ||
+                Request.IsAntaresInternalRequest())
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: Now returning detailed status info for Antares internal requests

File: src/WebJobs.Script/Diagnostics/HostPerformanceManager.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                return IsUnderHighLoad(counters);
+            }
+
+            return false;
+        }
+
+        internal static bool IsUnderHighLoad(ApplicationPerformanceCounters counters)
+        {
+            return
+                ThresholdExceeded(counters.Connections, counters.ConnectionLimit, ConnectionThreshold) ||
```
Comment (mathewc) [inline src/WebJobs.Script/Diagnostics/HostPerformanceManager.cs]: Simple heuristic for now that we'll expand on later

File: src/WebJobs.Script/Diagnostics/HostPerformanceManager.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            float currentUsage = (float)currentValue / limit;
+            return currentUsage > threshold;
+        }
+
+        internal ApplicationPerformanceCounters GetPerformanceCounters()
+        {
+            string json = _settingsManager.GetSetting(WebsiteAppCountersEnvironmentVariableName);
+            if (!string.IsNullOrEmpty(json))
+            {
+                // TEMP: need to parse this specially to work around bug where
```
Comment (mathewc) [inline src/WebJobs.Script/Diagnostics/HostPerformanceManager.cs]: Bilal has submitted a fix for this sandbox bug

File: src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs [hunk 5]
```diff
@@ -15,6 +16,19 @@ public static AuthorizationLevel GetAuthorizationLevel(this HttpRequestMessage r
...
+        {
+            if (!ScriptSettingsManager.Instance.IsAzureEnvironment)
+            {
+                return false;
+            }
+
+            // this header will *always* be present on requests originating externally (i.e. going
+            // through the Anatares front end). For requests originating internally it will NOT be
+            // present.
+            return !request.Headers.Contains(ScriptConstants.AntaresInternalRequestHeaderName);
```
Comment (mamaso) [inline src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs]: The naming of `AntaresInternalRequestHeaderName` is confusing, I think it should be `ExternalRequestHeaderName` or something along those lines, based on how it is used here

Comment (mathewc) [inline src/WebJobs.Script/Extensions/HttpRequestMessageExtensions.cs]: Changed

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs [hunk 6]
```diff
@@ -107,13 +110,22 @@ public HostStatus GetHostStatus()
                     status.Errors.Add(Utility.FlattenException(lastError));
                 }
 
-                return status;
+                var parameters = Request.GetQueryParameterDictionary();
+                string value = null;
+                if (parameters.TryGetValue(ScriptConstants.CheckLoadQueryParameterName, out value) && value == "1")
+                {
+                    status.Load = new LoadStatus
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: While we only have a simple bool to return now, I'm defining the payload to be extensible for the future. E.g. we'll likely want to return additional details in the future. To do so, we'll add members to LoadStatus.

File: src/WebJobs.Script/Config/ScriptSettingsManager.cs [hunk 7]
```diff
@@ -11,7 +11,8 @@ public class ScriptSettingsManager
         private static ScriptSettingsManager _instance = new ScriptSettingsManager();
         private readonly ConcurrentDictionary<string, string> _settingsCache = new ConcurrentDictionary<string, string>();
 
-        internal ScriptSettingsManager()
+        // for testing
+        public ScriptSettingsManager()
```
Comment (mathewc) [inline src/WebJobs.Script/Config/ScriptSettingsManager.cs]: Moq was failing to create a mock - making this public fixed it.

File: src/WebJobs.Script/Diagnostics/HostPerformanceManager.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.WebJobs.Script.Config;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.WebJobs.Script.Diagnostics
+{
+    public class HostPerformanceManager
+    {
+        internal const string WebsiteAppCountersEnvironmentVariableName = "WEBSITE_COUNTERS_APP";
```
Comment (mathewc) [inline src/WebJobs.Script/Diagnostics/HostPerformanceManager.cs]: Moved

**Pull request closed without merging**